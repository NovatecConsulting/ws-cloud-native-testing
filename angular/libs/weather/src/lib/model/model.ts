export interface WeatherLocation {
  temp: number;
  location: string;
}
