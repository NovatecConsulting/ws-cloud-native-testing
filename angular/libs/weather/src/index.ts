export * from './lib/weather.module';

export * from './lib/page-template/weather-template-data';
export * from './lib/model/model';
