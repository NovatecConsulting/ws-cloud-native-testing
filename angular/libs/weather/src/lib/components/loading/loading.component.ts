import { Component } from '@angular/core';

@Component({
  selector: 'weather-loading',
  templateUrl: './loading.component.html',
  styleUrls: ['./loading.component.scss'],
})
export class LoadingComponent {}
