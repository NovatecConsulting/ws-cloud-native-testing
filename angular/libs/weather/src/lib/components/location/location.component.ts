import { Component, EventEmitter, Input, Output } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';

@Component({
  selector: 'weather-location',
  templateUrl: './location.component.html',
  styleUrls: ['./location.component.scss'],
})
export class LocationComponent {
  static locationErrors = [
    { errorStatus: 'required', errorMessage: 'Please enter a location before searching!' },
    { errorStatus: 'minlength', errorMessage: 'At least 3 characters are needed!' },
    { errorStatus: 'pattern', errorMessage: 'Numbers are not allowed!' },
  ];
  @Input() defaultLocation: string | undefined = undefined;
  @Output() searched = new EventEmitter<string>();
  form: FormGroup = new FormGroup({
    location: new FormControl('', [Validators.required, Validators.pattern(new RegExp('^[a-zA-Z -]*$')), Validators.minLength(3)]),
  });

  startSearch() {
    if (this.form?.valid) {
      this.searched.emit(this.form.controls['location'].value);
    }
  }
}
