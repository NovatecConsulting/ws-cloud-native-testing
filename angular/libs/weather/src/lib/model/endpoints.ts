export class Endpoints {
  mainLocation = () => `${this.environment.weatherApi}/mainLocation`;
  queryLocation = (location: string) => `${this.environment.weatherApi}/locations?q=${location}`;

  constructor(private environment: {weatherApi: string}) {}
}
