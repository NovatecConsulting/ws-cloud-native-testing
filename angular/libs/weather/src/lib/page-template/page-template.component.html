<div class="block relative min-h-100">
  <ng-container *ngIf="(weatherTemplateData.isLoading$ | async) === false">
    <weather-location
      *ngIf="(weatherTemplateData.mainLocation$ | async) !== null"
      (searched)="searched.emit($event)"
      [defaultLocation]="(weatherTemplateData.mainLocation$ | async)!"
      class="block pb-4"
    ></weather-location>
    <weather-results
      *ngIf="weatherTemplateData.weather$ | async"
      [weatherLocations]="(weatherTemplateData.weather$ | async)!"
      (savedLocation)="savedLocation.emit($event)"
      class="block pb-4"
    ></weather-results>
    <weather-error-message *ngIf="weatherTemplateData.warning$ | async as message" [message]="message!" class="block mb-3"></weather-error-message>
  </ng-container>
  <weather-loading *ngIf="weatherTemplateData.isLoading$ | async"></weather-loading>
</div>
