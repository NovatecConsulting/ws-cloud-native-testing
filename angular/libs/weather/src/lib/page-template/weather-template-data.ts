import { Observable } from 'rxjs';
import { WeatherLocation } from '../model/model';

export type WeatherTemplateData = {
  weather$: Observable<WeatherLocation[]>;
  warning$: Observable<string | undefined>;
  isLoading$: Observable<boolean>;
  mainLocation$: Observable<string | undefined>;
};
