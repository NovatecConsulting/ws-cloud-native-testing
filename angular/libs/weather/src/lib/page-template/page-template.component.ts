import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { WeatherTemplateData } from '../page-template/weather-template-data';
import { WeatherLocation } from '../model/model';

@Component({
  selector: 'weather-page-template',
  templateUrl: './page-template.component.html',
  styleUrls: ['./page-template.component.scss'],
})
export class PageTemplateComponent implements OnInit {
  @Input() weatherTemplateData!: WeatherTemplateData;
  @Output() searched = new EventEmitter<string>();
  @Output() savedLocation = new EventEmitter<WeatherLocation>();

  ngOnInit(): void {
    if (!this.weatherTemplateData) {
      throw new Error('No data provided for weather template!');
    }
  }
}
