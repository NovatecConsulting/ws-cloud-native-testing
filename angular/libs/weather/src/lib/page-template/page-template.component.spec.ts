import { ComponentFixture, TestBed } from '@angular/core/testing';
import { HarnessLoader } from '@angular/cdk/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { ReactiveFormsModule } from '@angular/forms';
import { TestbedHarnessEnvironment } from '@angular/cdk/testing/testbed';
import { MatInputHarness } from '@angular/material/input/testing';
import { MockBuilder, MockRender } from 'ng-mocks';
import { MatCardModule } from '@angular/material/card';
import { MatExpansionModule } from '@angular/material/expansion';
import { MatInputModule } from '@angular/material/input';
import { MatButtonModule } from '@angular/material/button';
import { MatDividerModule } from '@angular/material/divider';
import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { PageTemplateComponent } from './page-template.component';
import { ErrorMessageComponent } from '../components/error-message/error-message.component';
import { LocationComponent } from '../components/location/location.component';
import { LoadingComponent } from '../components/loading/loading.component';
import { ResultsComponent } from '../components/result/results.component';
import { of } from 'rxjs';
import { MatButtonHarness } from '@angular/material/button/testing';
import { WeatherLocation } from '../model/model';
import { WeatherModule } from '../weather.module';
import { WeatherTemplateData } from './weather-template-data';

const MaterialModules = [MatCardModule, MatExpansionModule, MatInputModule, MatButtonModule, MatDividerModule, MatProgressSpinnerModule];

describe('A user visiting the Weather Page', () => {
  /**
   * These tests are testing the complete module and give us confidence for bigger refactorings.
   * They are rather brittle and thus, we have to use them carefully.
   * Further, they do not give good insights in why a specific scenario is failing.
   * This loads to a need of additional tests for your components to get better feedback.
   */
  describe('and having their data loaded', () => {
    let fixture: ComponentFixture<PageTemplateComponent>;
    let component: PageTemplateComponent;
    let loader: HarnessLoader;

    beforeEach(async () => {
      return TestBed.configureTestingModule({
        declarations: [PageTemplateComponent, ErrorMessageComponent, LocationComponent, LoadingComponent, ResultsComponent],
        imports: [NoopAnimationsModule, ReactiveFormsModule, ...MaterialModules],
      }).compileComponents();
    });

    beforeEach(() => {
      fixture = TestBed.createComponent(PageTemplateComponent);
      component = fixture.componentInstance;
      loader = TestbedHarnessEnvironment.loader(fixture);
    });

    it('should be able to search for a location', async () => {
      component.weatherTemplateData = {
        weather$: of([]),
        mainLocation$: of(undefined),
        isLoading$: of(false),
        warning$: of(undefined),
      };

      const inputLocation = 'Stuttgart';

      jest.spyOn(component.searched, 'emit');
      const locationInput = await loader.getHarness(MatInputHarness.with({ placeholder: 'Enter location' }));
      await locationInput.setValue(inputLocation);

      const searchButton = await loader.getHarness(MatButtonHarness.with({ text: 'Search' }));
      await searchButton.click();

      expect(component.searched.emit).toHaveBeenCalledWith(inputLocation);
    });

    it('should be able to save his location', async () => {
      const stuttgart = { location: 'Stuttgart', temp: 25 };
      const weather: WeatherLocation[] = [stuttgart];
      component.weatherTemplateData = {
        weather$: of(weather),
        mainLocation$: of(undefined),
        isLoading$: of(false),
        warning$: of(undefined),
      };

      jest.spyOn(component.savedLocation, 'emit');

      const saveButton = await loader.getHarness(MatButtonHarness.with({ text: 'Save Location & Reload' }));
      await saveButton.click();

      expect(component.savedLocation.emit).toHaveBeenCalledWith(stuttgart);
    });
  });

  describe('should have a consistent layout', () => {
    beforeEach(() => MockBuilder(PageTemplateComponent, WeatherModule));

    it('if their data is still loading', () => {
      const weatherTemplateData: WeatherTemplateData = {
        weather$: of([]),
        mainLocation$: of('Stuttgart'),
        isLoading$: of(true),
        warning$: of(undefined),
      };
      const fixture = MockRender(PageTemplateComponent, { weatherTemplateData });
      expect(fixture).toMatchSnapshot();
    });

    it('if their data contains an error', () => {
      const weatherTemplateData: WeatherTemplateData = {
        weather$: of([]),
        mainLocation$: of('Stuttgart'),
        isLoading$: of(false),
        warning$: of('Something went wrong!'),
      };
      const fixture = MockRender(PageTemplateComponent, { weatherTemplateData });
      expect(fixture).toMatchSnapshot();
    });

    it('if their data has valid results', () => {
      const weather: WeatherLocation[] = [
        { location: 'Stuttgart', temp: 25 },
        { location: 'Frankfurt', temp: 24 },
        { location: 'Berlin', temp: 22 },
      ];
      const weatherTemplateData: WeatherTemplateData = {
        weather$: of(weather),
        mainLocation$: of('Stuttgart'),
        isLoading$: of(false),
        warning$: of(undefined),
      };
      const fixture = MockRender(PageTemplateComponent, { weatherTemplateData });
      expect(fixture).toMatchSnapshot();
    });
  });
});
