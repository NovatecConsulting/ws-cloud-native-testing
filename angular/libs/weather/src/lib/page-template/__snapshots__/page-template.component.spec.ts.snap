// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A user visiting the Weather Page should have a consistent layout if their data contains an error 1`] = `
<mock-render
  __ngContext__={[Function Number]}
>
  <weather-page-template>
    <div
      class="block relative min-h-100"
    >
      <weather-location
        class="block pb-4"
      />
      <weather-results
        class="block pb-4"
      />
      <weather-error-message
        class="block mb-3"
      />
    </div>
  </weather-page-template>
</mock-render>
`;

exports[`A user visiting the Weather Page should have a consistent layout if their data has valid results 1`] = `
<mock-render
  __ngContext__={[Function Number]}
>
  <weather-page-template>
    <div
      class="block relative min-h-100"
    >
      <weather-location
        class="block pb-4"
      />
      <weather-results
        class="block pb-4"
      />
    </div>
  </weather-page-template>
</mock-render>
`;

exports[`A user visiting the Weather Page should have a consistent layout if their data is still loading 1`] = `
<mock-render
  __ngContext__={[Function Number]}
>
  <weather-page-template>
    <div
      class="block relative min-h-100"
    >
      <weather-loading />
    </div>
  </weather-page-template>
</mock-render>
`;
